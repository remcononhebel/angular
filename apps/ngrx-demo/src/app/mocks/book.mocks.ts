export const mockBooks = [
  {
    id: 'mock id 1',
    volumeInfo: { authors: ['mock authors 1'], title: 'mock title 1' },
  },
  {
    id: 'mock id 2',
    volumeInfo: { authors: ['mock authors 2'], title: 'mock title 2' },
  },
];
