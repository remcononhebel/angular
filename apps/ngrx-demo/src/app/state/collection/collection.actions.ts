import { createActionGroup, emptyProps, props } from '@ngrx/store';

export const CollectionActions = createActionGroup({
  source: 'Collection',
  events: {
    'Add Book': props<{ bookId: string }>(),
    'Remove Book': props<{ bookId: string }>(),
    'Select books': props<{ bookIds: string[] }>(),
    'Deselect all books': emptyProps()
  },
});

