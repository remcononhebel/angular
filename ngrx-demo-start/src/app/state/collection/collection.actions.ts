import { createActionGroup, props } from '@ngrx/store';
import { Book } from '../../model/book.model';

export const CollectionActions = createActionGroup({
  source: 'Collection',
  events: {
    'Add Book': props<{ bookId: string }>(),
    'Remove Book': props<{ bookId: string }>(),
    'Select books': props<{ bookIds: string[] }>()
  },
});

