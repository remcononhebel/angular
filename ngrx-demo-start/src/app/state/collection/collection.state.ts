
export interface CollectionState {
  selectedBooks: string[];
}
