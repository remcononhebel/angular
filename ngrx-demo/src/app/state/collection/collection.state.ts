
export interface CollectionState {
  books: string[];
}
