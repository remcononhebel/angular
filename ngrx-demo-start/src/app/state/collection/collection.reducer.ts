import { createReducer, on } from '@ngrx/store';
import { CollectionState } from './collection.state';
import { CollectionActions } from './collection.actions';

export const initialState: CollectionState = {
  selectedBooks: [],
};

export const collectionReducer = createReducer(
  initialState,
  on(CollectionActions.addBook, (state, { bookId }) => {
    if (state.selectedBooks.some((id) => id === bookId)) {
      return state;
    }

    return {
      ...state,
      selectedBooks: [...state.selectedBooks, bookId],
    };
  }),
  on(CollectionActions.removeBook, (state, { bookId }) => ({
    ...state,
    selectedBooks: state.selectedBooks.filter((id) => id !== bookId),
  })),
  on(CollectionActions.selectBooks, (state, { bookIds }) => ({
    ...state,
    selectedBooks: [...bookIds],
  }))
);
