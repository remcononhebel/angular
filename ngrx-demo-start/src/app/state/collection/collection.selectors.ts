import { createFeatureSelector, createSelector } from '@ngrx/store';
import { Book } from '../../model/book.model';
import { selectBooksState, selectNumberOfAvailableBooks } from '../books/books.selectors';
import { CollectionState } from './collection.state';

export const selectCollectionState = createFeatureSelector<CollectionState>('collection');

export const selectBookCollection = createSelector(selectBooksState, selectCollectionState, (booksState, collectionState) => {
  return collectionState.books.map((id) => booksState.books.find((book) => book.id === id)!);
});

export const selectNumberOfBooksInCollection = createSelector(
  selectNumberOfAvailableBooks,
  selectBookCollection,
  (numberOfAvailableBooks: number, booksInCollection: Book[]) => {
    if (booksInCollection.length === 0) {
      return 'No books selected';
    }

    if (booksInCollection.length === numberOfAvailableBooks) {
      return 'All available books selected';
    }
    return `${booksInCollection.length} / ${numberOfAvailableBooks} books selected`;
  }
);
