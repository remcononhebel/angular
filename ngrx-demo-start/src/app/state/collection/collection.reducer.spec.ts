import { mockBooks } from '../../mocks/book.mocks';
import { CollectionActions } from './collection.actions';
import { collectionReducer } from './collection.reducer';
import { CollectionState } from './collection.state';

describe('collectionReducer', () => {
  describe('addBook', () => {
    it('should add a book', () => {
      const state: CollectionState = {
        selectedBooks: [],
      };
      expect(collectionReducer(state, CollectionActions.addBook({ bookId: 'mock book id' }))).toMatchSnapshot();
    });

    it('should not add a book when it is already in the state', () => {
      const bookId = 'mock book id';
      const state: CollectionState = {
        selectedBooks: [bookId],
      };
      expect(collectionReducer(state, CollectionActions.addBook({ bookId }))).toMatchSnapshot();
    });
  });

  describe('removeBook', () => {
    it('should remove a book that is in the state', () => {
      const bookId = 'mock book id';
      const state: CollectionState = {
        selectedBooks: [bookId],
      };
      expect(collectionReducer(state, CollectionActions.removeBook({ bookId }))).toMatchSnapshot();
    });

    it('should not remove a book that is not in the state', () => {
      const bookId = 'mock book id';
      const state: CollectionState = {
        selectedBooks: [bookId],
      };
      expect(collectionReducer(state, CollectionActions.removeBook({ bookId: 'does not exist' }))).toMatchSnapshot();
    });
  });

  describe('selectBooks', () => {
    it('should select multiple books', () => {
      const state: CollectionState = { selectedBooks: [] };
      expect(collectionReducer(state, CollectionActions.selectBooks({bookIds: mockBooks.map(({id}) => id)}))).toMatchSnapshot();
    });
  });
});
