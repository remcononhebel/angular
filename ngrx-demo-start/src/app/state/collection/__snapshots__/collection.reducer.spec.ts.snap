// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectionReducer addBook should add a book 1`] = `
{
  "selectedBooks": [
    "mock book id",
  ],
}
`;

exports[`collectionReducer addBook should not add a book when it is already in the state 1`] = `
{
  "selectedBooks": [
    "mock book id",
  ],
}
`;

exports[`collectionReducer removeBook should not remove a book that is not in the state 1`] = `
{
  "selectedBooks": [
    "mock book id",
  ],
}
`;

exports[`collectionReducer removeBook should remove a book that is in the state 1`] = `
{
  "selectedBooks": [],
}
`;

exports[`collectionReducer selectBooks should select multiple books 1`] = `
{
  "selectedBooks": [
    "mock id 1",
    "mock id 2",
  ],
}
`;
